### DISCUSSION TITLE
How to include two versions of the same port in a project?

### DISCUSSION THREAD

**Discussion by zoharl3** [microsoft/vcpkg#41793] [Q&A]
I have a project with two cmake targets:
1. A static lib that requires cgal 5.
2. The main exe that requires cgal 6.
How do I accomplish that with vcpkg?

**Comment by autoantwort** [comment]
That is not possible since the header files would conflict.

**Reply by zoharl3** [reply]
Here, look at this:

https://stackoverflow.com/questions/58316523/how-to-include-two-different-versions-of-the-same-dependency-in-a-cmake-project

cmake supports it; does vcpkg?

**Reply by Osyotr** [reply]
CMake supports **finding** different versions of the same package (with a handful of limitations).
Vcpkg installs all packages into the same folder, so **installing** two versions of the same package would result in a file conflict. There are exceptions like qt5 and qt6, but they explicitly support this configuration. I don't think cgal supports that.
Note that linking two versions of the same package this is usually a very bad thing to do.

**Reply by zoharl3** [reply]
Maybe installing all packages into the same folder isn't such a great idea. What if two packages have the same name of a .lib file?

It's not ideal, and I'm not doing it for the fun of it. But when there are breaking changes, it's unavoidable. If there's collision in the linkage, then the targets would have to be dynamic rather than static. In the case of CGAL, it's header only.

**Reply by dg0yt** [reply]
Even if the libs have different names or are "header-only", there is a minefield of ABI issues.

**Reply by zoharl3** [reply]
- It needs to be done.
- I can do it manually.
- It would be nice if vcpkg supported it.

**Reply by autoantwort** [reply]
If I understand the SO answer right you already have two separate CMakeLists.txt. So you could create two separate vcpkg.json files and configure and build each CMakeLists.txt separately. (I assume your main CMakeLists.txt file is just
```cmake
project(MyProject)
add_subdirectory(folder_one)
add_subdirectory(folder_two)
```

**Reply by zoharl3** [reply]
(SO--significant other?)

Yes, this is my main cmake list.

In folder1, I have a manifest
```
{
  "dependencies": [
    { "name": "cgal" }
  ]
}
```
In folder2, I have a manifest
```
{
  "dependencies": [
    { "name": "cgal" }
  ],
  "builtin-baseline": "edc84ff66e6262a9f7565c28eb76577aeab9c5aa",
  "overrides": [
    {
      "name": "cgal",
      "version": "5.6.1"
    }
  ]
}
```
Both folders have a cmakefile
```
find_package(CGAL COMPONENTS Core REQUIRED)
```
I get an error that it can't find gmp.

**Reply by autoantwort** [reply]
SO=Stackoverflow 

> I get an error that it can't find gmp.

Can't reproduce. I used your vcpkg.json files, the following CMakeLists.txt file:
```cmake
cmake_minimum_required(VERSION 3.21)
project(test)
find_package(CGAL COMPONENTS Core REQUIRED)
```
and called `cmake -S . -B build -GNinja -DCMAKE_TOOLCHAIN_FILE="vcpkg/scripts/buildsystems/vcpkg.cmake"` and it worked.

**Reply by zoharl3** [reply]
You were supposed to call find within each subdir. See attachment.
[cpp2.zip](https://github.com/user-attachments/files/17539515/cpp2.zip)

**Reply by Osyotr** [reply]
You've missed the crucial step:
> configure and build each CMakeLists.txt separately

Otherwise, it's not possible.

**Reply by zoharl3** [reply]
You mean three (two) independent projects; that's one option but not what I'm looking for.

It would be nice to have such flexibility.
